src/napari_graph/_base_graph.py

```diff
@@ -0,0 +1,749 @@
+    n_nodes : int
+        Optional number of nodes to pre-allocate in the graph.
+    n_edges : int
+        Optional number of edges to pre-allocate in the graph.
+    """
+
+    _NODE_EMPTY_PTR = -1
+
+    # abstract constants
+    _EDGE_DUPLICATION: int = ...
+    _EDGE_SIZE: int = ...
+    _LL_EDGE_POS: int = ...
+
+    # allocation constants
+    _ALLOC_MULTIPLIER = 1.1
+    _ALLOC_MIN = 25
+
+    def __init__(
+        self,
+        edges: ArrayLike = [],
+        coords: Optional[Union[pd.DataFrame, ArrayLike]] = None,
+        dim: Optional[int] = None,
+        n_nodes: Optional[int] = None,
+        n_edges: Optional[int] = None,
+    ):
+
+        # validating edges length
+        if n_edges is None:
+            n_edges = len(edges)
+
+        if n_edges < len(edges):
+            raise ValueError(
+                f"`n_edges` ({n_edges}) must be greater or equal than `edges` length ({len(edges)}."
+            )
+
+        # validating n_nodes and coords
+        self._coords = None
+        if coords is not None:
+            if dim is not None:
+                raise ValueError(
+                    "`dim` and `coords` cannot be supplied at the same time."
+                )
+
+            if not isinstance(coords, pd.DataFrame):
+                coords = np.asarray(coords)
+
+            dim = coords.shape[1]
+
+            if n_nodes is None:
+                n_nodes = len(coords)
+
+            if n_nodes < len(coords):
+                raise ValueError(
+                    f"`n_nodes` ({n_nodes}) must be greater or equal than `coords` length ({len(coords)})."
+                )
+
+        elif n_nodes is None:
+            n_nodes = 0
+
+        # allocates coords if dim was provided
+        if dim is not None:
+            self._coords = np.empty((n_nodes, dim), dtype=np.float32)
+
+        self._init_buffers(n_nodes=n_nodes, n_edges=n_edges)
+
+        if coords is not None:
+            self.init_data_from_dataframe(coords)
+
+        if len(edges) > 0:
+            self.add_edges(edges)
+
+    def _init_buffers(self, n_nodes: int, n_edges: int) -> None:
+        # node-wise buffers
+        self._empty_nodes: List[int] = list(reversed(range(n_nodes)))
+        self._node2edges = np.full(
+            n_nodes, fill_value=_EDGE_EMPTY_PTR, dtype=int
+        )
+        self._world2buffer = typed.Dict.empty(types.int64, types.int64)
+        self._buffer2world = np.full(
+            n_nodes, fill_value=self._NODE_EMPTY_PTR, dtype=int
+        )
+        # edge-wise buffers
+        self._empty_edge_idx = 0 if n_edges > 0 else _EDGE_EMPTY_PTR
+        self._n_edges = 0
+        self._edges_buffer = np.full(
+            n_edges * self._EDGE_DUPLICATION * self._EDGE_SIZE,
+            fill_value=_EDGE_EMPTY_PTR,
+            dtype=int,
+        )
+        self._edges_buffer[
+            self._LL_EDGE_POS : -self._EDGE_SIZE : self._EDGE_SIZE
+        ] = np.arange(1, self._EDGE_DUPLICATION * n_edges)
```

Comment:
> Maybe split up into `_init_node_buffers` and `_init_edge_buffers`? Happy to submit a PR if desired.

See https://github.com/JoOkuma/napari-graph/pull/9

> Also, maybe rename as follows (mainly for consistency)?

For this, I'll wait for feedback from you & for the PR above

Reply:
Hi @JoOkuma 

Re the renaming: the reason I suggested this was to make the naming of variables a bit more consistent, but this isn't restricted to just this part of the code. There are a number of places in the current implementation where it isn't instantly clear whether e.g. an *edge* is supposed to be a 2-tuple, an *edge number*, or an *edge buffer index*. Same goes for *node*, where it e.g. isn't clear from the current naming scheme that this in many cases is a *node index* and not a *node label*.

Trying to come up with a more generalized terminology, I propose the following naming scheme (hopefully capturing most of the information floating around):
- *node index*, i.e. an integer in the range `[0, n_nodes)` (also directly serves as *node buffer index*)
- *node label*, i.e. a user-defined unique node identifier (mapped to the corresponding *node index* via `world2buffer`/`buffer2world` arrays in the current implementation, if I understand correctly)
- *node coordinate*, i.e. a `d`-tuple of floats (mapped via *node index*)
- *edge*, i.e. a unique 2-tuple of *node indices* (not *node labels* (!) if I understand correctly)
- *edge number*, i.e. an integer in the range `[0, n_edges)`
- *edge buffer index*, i.e. an integer in the range `[0, len(edges_buffer))`

... and to refactor the code accordingly to make the names of attributes, arguments, etc a bit more clear/expressive/consistent. What do you think?

Happy to give this a try, but it may turn out to be quite a "busy" PR I guess :-)